packages/deposit_firebase/lib/src/deposit_firebase_adapter.dart

```diff
@@ -0,0 +1,145 @@
+import 'package:cloud_firestore/cloud_firestore.dart';
+import 'package:deposit/deposit.dart';
+
+extension _QueryDocumentSnapshotX on DocumentSnapshot {
+  Map<String, dynamic> toMap() {
+    final _data = (data() ?? {}) as Map<String, dynamic>;
+
+    return <String, dynamic>{
+      'id': id,
+      ..._data,
+    };
+  }
+}
+
+class FirebaseDepositAdapter extends DepositAdapter<String> {
+  FirebaseDepositAdapter({
+    required FirebaseFirestore firestore,
+  }) : _firestore = firestore;
+
+  final FirebaseFirestore _firestore;
+
+  @override
+  Future<Map<String, dynamic>> add(
+    String table,
+    String primaryColumn,
+    Map<String, dynamic> data,
+  ) async {
+    final value = await _firestore.collection(table).add(data);
+
+    final snapshot = await value.snapshots().first;
+    final returnedData = snapshot.data();
+
+    if (returnedData == null) {
+      throw Exception('Unable to insert data into "$table"');
+    } else {
+      return snapshot.toMap();
+    }
+  }
+
+  @override
+  Future<List<Map<String, dynamic>>> by(String table, String key, value) async {
+    final data = await _firestore
+        .collection(table)
+        .where(
+          key,
+          isEqualTo: value,
+        )
+        .get();
+
+    return data.docs.map((doc) => doc.toMap()).toList();
+  }
+
+  @override
+  Future<bool> exists(String table, String primaryColumn, String id) async {
+    final snap = await _getDoc(primaryColumn, table, id);
+    return snap?.exists ?? false;
+  }
+
+  @override
+  Future<Map<String, dynamic>> getById(
+    String table,
+    String primaryColumn,
+    String id,
+  ) async {
+    final snap = await _getDoc(primaryColumn, table, id);
+
+    if (snap == null) {
+      /// Probably better to change this method signature to allow null
+      throw Exception('Unable to find doc');
+    }
+
+    return snap.toMap();
+  }
+
+  @override
+  Future<List<Map<String, dynamic>>> page(
+    String table, {
+    required int limit,
+    required int skip,
+    OrderBy? orderBy,
+  }) async {
+    throw UnsupportedError(
+      "Firebase adapter can't paginate with numeric offset",
+    );
+  }
+
+  @override
+  Future<void> remove(
+    String table,
+    String primaryColumn,
+    Map<String, dynamic> data,
+  ) async {
+    final doc = await _getDocRef(primaryColumn, table, data[primaryColumn]);
```

Comment:
```suggestion
    // TODO: Check if the id is not null and throw exception in case it is
    final doc = await _getDocRef(primaryColumn, table, data[primaryColumn]);
```